crates/supervisor/storage/src/providers/derivation_provider.rs

```diff
@@ -0,0 +1,181 @@
+//! Provider for derivation-related database operations.
+
+use crate::{error::StorageError, models::{self, DerivedBlocks, StoredDerivedBlockPair}};
+use alloy_eips::eip1898::BlockNumHash;
+use kona_interop::DerivedRefPair;
+use kona_protocol::BlockInfo;
+use reth_db_api::{
+  cursor::{DbCursorRO, DbDupCursorRO, DbDupCursorRW},
+  transaction::{DbTx, DbTxMut}
+};
+use tracing::{error, warn};
+
+/// Provides access to derivation storage operations within a transaction.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub(crate) struct DerivationProvider<'tx, TX> {
+    tx: &'tx TX,
+}
+
+impl<'tx, TX> DerivationProvider<'tx, TX> {
+    /// Creates a new [`DerivationProvider`] instance.
+    #[allow(dead_code)]
+    pub(crate) const fn new(tx: &'tx TX) -> Self {
+        Self { tx }
+    }
+}
+
+impl<TX> DerivationProvider<'_, TX>
+where
+    TX: DbTx,
+{
+    /// Helper to get [`StoredDerivedBlockPair`] by block number.
+    #[allow(dead_code)]
```

Comment:
```suggestion
    /// Helper to get [`StoredDerivedBlockPair`] by block number.
    #[expect(dead_code)]
```
expect attribute makes sure a warning is emitted by compiler when this code is not dead code, so we know when we can remove the attribute